performance.py

```diff
@@ -558,12 +608,45 @@ async def main(
             + timedelta(seconds=running_time),
         )
 
+        def event_consumer(event: "yapapi.events.AgreementTerminated"):
+            fail_provider = event.agreement.details.provider_node_info.name
+            # Check if Provider is already listed as the test node
+            if fail_provider in mapping.values():
+                for ip_fail_provider, name_fail_provider in ip_provider_name.items():
+                    if name_fail_provider == fail_provider:
+                        network_addresses.remove(ip_fail_provider)
+                        computation_state_client[ip_fail_provider] = State.FINISHED
+                        for ip, completed_set in completion_state.items():
+                            completion_counter[ip] = completion_counter[ip] + 1
+                            if ip_fail_provider in completed_set:
+                                completed_set.remove(ip_fail_provider)
+                        if ip_fail_provider in current_computation.keys():
+                            if (
+                                computation_state_client[current_computation[ip_fail_provider]]
+                                != State.FINISHED
+                            ):
+                                computation_state_client[
+                                    current_computation[ip_fail_provider]
+                                ] = State.IDLE
+                        if ip_fail_provider in current_computation.values():
+                            key_list = list(current_computation.keys())
+                            val_list = list(current_computation.values())
+                            server_ip = key_list[val_list.index(ip_fail_provider)]
+                            computation_state_server[server_ip] = State.IDLE
```

Comment:
I would say, that your life would be much easier, if you would encapsulate all these state related globals in one or few objects managing state. Than you could create methods for common state changes.
If you have functions with meaningful names, than it is easier to read and develop code, because these names add semantic to code and hide complexity.

But it is only comment for future, not for now

Reply:
You're totally right! I think that also some kind of round robin algorithm may help get rid of those locks in the code, but I'll leave it for later.